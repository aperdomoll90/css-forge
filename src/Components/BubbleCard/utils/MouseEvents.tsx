export const rotateElement = (event: MouseEvent, element: Element | null) => {
  // get mouse position
  const x = event.clientX
  const y = event.clientY
  // find the middle
  const middleX = window.innerWidth / 2
  const middleY = window.innerHeight / 2

  //get offset from middle
  const offsetX = ((x - middleX) / middleX) * 25
  const offsetY = ((y - middleY) / middleY) * 25
  console.log('element', element)
  if (element) {
    element.style.setProperty('--rotateX', -1 * offsetY + 'deg')
    element.style.setProperty('--rotateY', offsetX + 'deg')
  }
}
