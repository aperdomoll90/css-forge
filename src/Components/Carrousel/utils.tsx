import * as React from 'react'


function getPrevElement(list:any) {
    const sibling = list[0].previousElementSibling
  
    if (sibling instanceof HTMLElement) {
      return sibling
    }
  
    return sibling
  }
  
  function getNextElement(list:any) {
    const sibling = list[list.length - 1].nextElementSibling
  
    if (sibling instanceof HTMLElement) {
      return sibling
    }
  
    return null
  }

  export default function usePosition(ref) {
    const [prevElement, setPrevElement] = React.useState(null)
    const [nextElement, setNextElement] = React.useState(null)
  
    React.useEffect(() => {
      const element = ref.current
  
      const update = () => {
        const rect = element.getBoundingClientRect()
  
        const visibleElements:any  = Array.from(element.children).filter(child => {
          const childRect = child.getBoundingClientRect()
  
          return childRect.left >= rect.left && childRect.right <= rect.right
        })
  
        if (visibleElements.length > 0) {
          setPrevElement(getPrevElement(visibleElements))
          setNextElement(getNextElement(visibleElements))
        }
      }
  
      update()
  
      element.addEventListener('scroll', update, { passive: true })
  
      return () => {
        element.removeEventListener('scroll', update, { passive: true })
      }
    }, [ref])
  
    const scrollToElement = React.useCallback(
      element => {
        const currentNode = ref.current
  
        if (!currentNode || !element) return
  
        let newScrollPosition
  
        newScrollPosition =
          element.offsetLeft +
          element.getBoundingClientRect().width / 2 -
          currentNode.getBoundingClientRect().width / 2
  
        currentNode.scroll({
          left: newScrollPosition,
          behavior: 'smooth',
        })
      },
      [ref]
    )
  
    const scrollRight = React.useCallback(
      () => scrollToElement(nextElement),
      [scrollToElement, nextElement]
    )
  
    const scrollLeft = React.useCallback(
      () => scrollToElement(prevElement),
      [scrollToElement, prevElement]
    )
  
    return {
      hasItemsOnLeft: prevElement !== null,
      hasItemsOnRight: nextElement !== null,
      scrollRight,
      scrollLeft,
    }
  }