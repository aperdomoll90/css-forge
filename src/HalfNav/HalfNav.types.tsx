import { menuItemsArrayPropsTypes } from '../utils/GlobalTypes.types'

export interface HalfNavPropsType {
  menuItemsArray: menuItemsArrayPropsTypes[]
  color?: `#${string}` | `rgb(${number}, ${number}, ${number})` | `rgba(${number}, ${number}, ${number}, ${number})` | ''
  buttonHover?: string
  labelColor?: string
  labelColorHover?: string
  buttonBackgroundColor?: string
  menuBackgroundColor?: string
}
