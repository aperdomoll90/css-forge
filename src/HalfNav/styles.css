.HalfNavWrapper {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  position: relative;
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
  overflow: hidden;
}

.HalfNavDropDown {
  width: 40%;
  height: 100%;
  padding: 5rem 2rem;
  position: absolute;
  top: 0;
  left: -40%;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
  background: var(--menuBackgroundColor);
  box-shadow: 0 10px 10px rgba(0, 0, 0, 0.5);
  z-index: 0;
  opacity: 0;
  transition: 0.9s ease-in-out;
  transition-delay: 0.25s;
}
.HalfNavDropDown-menu-bullet {
  height: 2rem;
  width: 10rem;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
}

.HalfNavDropDown-menu-bullet svg {
  height: 80%;
}
.HalfNavDropDown-menu-bullet svg path {
  stroke-width: 1;
  stroke: yellow;
}

.HalfNavDropDown p {
  color: var(--labelColor);
  font-weight: 500;
  font-size: 1.2rem;
  letter-spacing: 4px;
  transition: 0.5s ease-in-out;
  cursor: pointer;
}
.HalfNavDropDown p:hover {
  text-shadow: 0px 3px 10px black;
  color: var(--labelColorHover);
}

.HalfNavDropDownActive {
  transform: translateX(100%);
  opacity: 1;
}
