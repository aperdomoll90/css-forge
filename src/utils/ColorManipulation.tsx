import { useEffect } from 'react'
import { useInView } from 'react-intersection-observer'

export interface ColorPropsType {
  color: string | `#${string}` | `rgb(${number}, ${number}, ${number})` | `rgba(${number}, ${number}, ${number}, ${number})`
}

export const RandomColor = (): string => { 
  const randomColor = Math.floor(Math.random()*16777215).toString(16)
  return `#${randomColor}`
}

export const ColorByIndex = (index:number, arraySize:number) => { 
  return `hsl(${(index * 360) / arraySize}, 100%, 50%)`
}

export const ChangeSectionColor = (bgColor: string, inView: boolean, parentID: string): void => {
  // add this where this function will be used
  // const [refItem, inView] = useInView({
  //   threshold: 0,
  // })
  const wrapper = document.getElementById(parentID)
  useEffect(() => {
    inView && wrapper && (wrapper.style.backgroundColor = bgColor)
  }, [inView])
}

export const LightenDarkenColor = (color: any, amount: number) => {
  var usePound = false
  if (color[0] == '#') {
    color = color.slice(1)
    usePound = true
  }

  var num = parseInt(color, 16)

  var r = (num >> 16) + amount

  if (r > 255) r = 255
  else if (r < 0) r = 0

  var b = ((num >> 8) & 0x00ff) + amount

  if (b > 255) b = 255
  else if (b < 0) b = 0

  var g = (num & 0x0000ff) + amount

  if (g > 255) g = 255
  else if (g < 0) g = 0

  return (usePound ? '#' : '') + (g | (b << 8) | (r << 16)).toString(16)
}
