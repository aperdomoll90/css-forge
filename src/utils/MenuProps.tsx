import { menuItemsArrayPropsTypes } from './GlobalTypes.types'
import { BarIcon, CameraIcon, CodeIcon, GearIcon, HomeIcon, MapIcon, PawIcon, PlanetIcon, ProfileIcon } from '../IconCollection/IconCollection'
export const menuItemsArray: menuItemsArrayPropsTypes[] = [
  {
    label: 'Home',
    icon: HomeIcon,
    link: '#',
  },
  {
    label: 'Settings',
    icon: GearIcon,
    link: '#',
  },
  {
    label: 'Data',
    icon: BarIcon,
    link: '#',
  },
  {
    label: 'Profile',
    icon: ProfileIcon,
    link: '#',
  },
  {
    label: 'Code',
    icon: CodeIcon,
    link: '#',
  },
]

export const extendedMenuItemsArray: menuItemsArrayPropsTypes[] = [
    ...menuItemsArray,
    {
      label: 'Map',
      icon: MapIcon,
      link: '#',
    },
    {
      label: 'Animations',
      icon: CameraIcon,
      link: '#',
    },
    {
      label: 'Pets',
      icon: PawIcon,
      link: '#',
    },
  ]