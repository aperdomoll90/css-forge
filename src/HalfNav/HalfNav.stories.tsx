import { Meta, StoryFn } from '@storybook/react'
import { HalfNav } from './HalfNav'
import StorybookContainer from '../StorybookContainer'
import { HalfNavPropsType } from './HalfNav.types'
import { menuItemsArray } from '../utils/MenuProps'

export default {
  title: 'Navigation Bars',
  component: HalfNav,
  parameters: {
    jest: ['HalfNav.test.tsx'],
  },
  argTypes: {
    color: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
    buttonHover: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
    labelColor: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
    labelColorHover: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
    buttonBackgroundColor: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
    menuBackgroundColor: { options: ['', 'black', 'white', 'red', 'blue'], control: { type: 'radio' } },
  },
} as Meta

const HalfNavTemplate: StoryFn<HalfNavPropsType> = args => (
  <StorybookContainer>
    <div className='story-button-container-centered'>
      <HalfNav {...args} />
    </div>
  </StorybookContainer>
)

export const HalfNavComponent = HalfNavTemplate.bind({})
HalfNavComponent.args = {
  menuItemsArray: menuItemsArray,
}
