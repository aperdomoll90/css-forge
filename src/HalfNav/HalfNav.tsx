import { useEffect, useState } from 'react'
import './styles.css'
import '../utils/GlobalStyles.css'
import { HalfNavPropsType } from './HalfNav.types'
import { menuItemsArrayPropsTypes } from '../utils/GlobalTypes.types'
import { LightenDarkenColor } from '../utils/ColorManipulation'
import { ToggleButton } from '../ToggleButton'

export const HalfNav: React.FC<HalfNavPropsType> = ({ menuItemsArray, color, buttonHover, labelColor, labelColorHover, buttonBackgroundColor, menuBackgroundColor }) => {
  const [showMenu, setShowMenu] = useState<boolean>(false)

  useEffect(() => {
    console.log('showMenu', showMenu)
  }, [showMenu])

  const defaultColor = '#fff'
  const defaultLabelColor = '#fff'
  const defaultLabelColorHover = '#7a7777'
  const defaultBackgroundColor = '#303030da'

  const stylesProps = {
    '--color': color ? color : defaultColor,
    '--buttonHover': buttonHover ? buttonHover : LightenDarkenColor(defaultBackgroundColor, -80),
    '--labelColor': labelColor ? labelColor : defaultLabelColor,
    '--labelColorHover': labelColorHover ? labelColorHover : defaultLabelColorHover,
    '--menuBackgroundColor': menuBackgroundColor ? menuBackgroundColor : defaultBackgroundColor,
    '--buttonBackgroundColor': buttonBackgroundColor ? buttonBackgroundColor : defaultBackgroundColor,
  }

  useEffect(() => {
    let menuDropDown = document.querySelector('.HalfNavDropDown') as HTMLElement
    showMenu ? menuDropDown.classList.add('HalfNavDropDownActive') : menuDropDown.classList.remove('HalfNavDropDownActive')
  }, [showMenu])

  return (
    <div className='HalfNavWrapper' style={stylesProps as React.CSSProperties}>
      <ToggleButton setActive={setShowMenu} color={color} buttonHover={buttonHover} buttonBackgroundColor={buttonBackgroundColor} />
      <div className='HalfNavDropDown'>
        {menuItemsArray &&
          menuItemsArray.map((item: menuItemsArrayPropsTypes, index: number) => {
            return (
              <div className='HalfNavDropDown-menu-bullet'>
                <p>{item.label}</p>
              </div>
            )
          })}
      </div>
    </div>
  )
}
