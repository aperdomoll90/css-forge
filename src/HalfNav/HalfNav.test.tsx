import { render } from '@testing-library/react'
import {HalfNav} from './HalfNav'
import { menuItemsArray } from '../utils/MenuProps'

describe('HalfNav component', () => {
const props = {
  menuItemsArray: menuItemsArray,
}
  const renderComponent = () => render(<HalfNav {...props} />)
  it('should match snapshots', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
