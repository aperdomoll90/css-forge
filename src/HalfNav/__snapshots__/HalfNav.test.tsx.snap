// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HalfNav component should match snapshots 1`] = `
<div>
  <div
    class="HalfNavWrapper"
    style="--color: #fff; --buttonHover: #ff008a; --labelColor: #fff; --labelColorHover: #7a7777; --menuBackgroundColor: #303030da; --buttonBackgroundColor: #303030da;"
  >
    <button
      class="ToggleButton"
      style="--size: 2rem; --color: #fff; --buttonHover: #ff008a; --buttonBackgroundColor: #303030da; --top: 1rem; --bottom: auto; --left: auto; --right: 1rem;"
    >
      <span />
      <span />
      <span />
      <span
        class="sr-only"
      >
        menu
      </span>
    </button>
    <div
      class="HalfNavDropDown"
    >
      <div
        class="HalfNavDropDown-menu-bullet"
      >
        <p>
          Home
        </p>
      </div>
      <div
        class="HalfNavDropDown-menu-bullet"
      >
        <p>
          Settings
        </p>
      </div>
      <div
        class="HalfNavDropDown-menu-bullet"
      >
        <p>
          Data
        </p>
      </div>
      <div
        class="HalfNavDropDown-menu-bullet"
      >
        <p>
          Profile
        </p>
      </div>
      <div
        class="HalfNavDropDown-menu-bullet"
      >
        <p>
          Code
        </p>
      </div>
    </div>
  </div>
</div>
`;
